### ISSUE TITLE
Set and Get PATCHING Routes

### ISSUE THREAD

**Issue by Suenee** [bitfocus/companion-module-soundcraft-ui#100]
**SET PATACHING ROUTE**

For one project, it would be very useful if it were possible to change the path settings in PATCHING. Specifically, we would need to change the sound source at the headphone output. Would it be possible to program this please?

![Image](https://github.com/user-attachments/assets/c020f229-014c-48f7-9e4e-5ab62f16be9e)
I suppose you could create a command that has four select boxes in it:
1. specifies output interfaces: CHANNELS, HW OUTS, CASCADE OUTS
2. determine the input interface: HW INS, USB-A, USB-DAW, CASCADE INS, MASTERS
3. addresses the route of the internal source (route from)
4. addresses the path where the route leads (route to)

**Feedback GET PATCHING ROUTE**
Verification that the given route is set on the patch page.

____
OFFTOPIC: Please send me your account number in the international format, I will be happy to support you also for that modification for MULTITRACK RECORD as I promised. MTR works well! :-)

**Comment by fmalcher** [comment]
Thanks for your feature request! This looks pretty complex if done right so I'm a bit hesitant to fully implement this.
Is it a solution for you to route an AUX bus to the headphone output and mute/unmute channels on this AUX mix?

> OFFTOPIC: Please send me your account number in the international format, I will be happy to support you 

That's great, thank you very much! The easiest way is to use GitHub Sponsors: https://github.com/sponsors/fmalcher/

**Comment by Suenee** [comment]
We do a big conference production once a year (it was 14 days ago now) where we cascade two SoundCrafts. One is in the control room (master) and the other is on the stage (slave). Two sound engineers are in charge of 6 microphones, 2 sound inputs from computers and one zoom call. They mix the outputs into two live broadcasts (one in Czech and the other in English). (You can tell by looking at the picture.)

We are looking for an easy solution so that the sound engineers (everyone to himself) could choose what to listen to in the headphones on the output of HEARDPHONE 1 and 2.

We just thought of using routing in the PATCHING table. Through the Companion, we would switch which internal channel will go to the H1 and/or H2 output.

If it makes your job easier, I suggest limiting myself to the page where HW outputs are routed to ports. When it comes to eavesdropping on AUXes, that is very important. (On the AUX outputs they listen to the sound for the language mutations of the broadcast).

Other combinations are not so important to us.

Hopefully it could be simplified like this. Is that clear please?

**Comment by fmalcher** [comment]
## Action
![Image](https://github.com/user-attachments/assets/9737db0e-2c1a-4ce8-a24c-71abb8ec0953)

## Feedback
![Image](https://github.com/user-attachments/assets/b152ba20-1086-456c-b696-72b82018d740)

## Source selection
![Image](https://github.com/user-attachments/assets/7490d724-45c6-49c7-8060-b39f343e937a)

## Destination selection
![Image](https://github.com/user-attachments/assets/48115caf-e8ec-465a-af60-0e314e28c6c7)

**Comment by fmalcher** [comment]
@Suenee This is my first idea for this. To keep things simple, I decided to create ONE dropdown for each source and destination so that we don't have to validate the input value. Since patching is only possible on the Ui24R, we know the number of inputs and outputs.
I tested a few scenarios and I'm confident it will serve exactly your use case. What do you think?

**Comment by Suenee** [comment]
> Here's my first idea for this. To keep things simple, I decided to create one dropdown each for source and destination so that we don't have to validate the numeric input value.
Sure, I agree, that dropdown is fine. It's just not clear to me from the picture, does SOURCE (FROM) and DESTIONATION (TO) only offer combinations that are feasible? I'm just asking to be sure.

> Since patching is only possible on the Ui24R, we know the exact number of inputs and outputs.
Great

> I tested a few scenarios and I'm confident it will serve exactly your use case. What do you think?
As you describe it, it looks great. I would love to help you and try it out too. Is it possible to download the changed files somewhere and temporarily overwrite production version with the newest? Or how best to proceed?

**Comment by fmalcher** [comment]
> It's just not clear to me from the picture, does SOURCE (FROM) and DESTIONATION (TO) only offer combinations that are feasible? I'm just asking to be sure.

The only combination that I know doesn't work is patching USB-A inputs to HW OUTS or CASCASE OUTS. Do you know any other impossible combinations?

> Is it possible to download the changed files somewhere and temporarily overwrite production version with the newest? Or how best to proceed?

Unfortunately not really an easy way. You will have to clone the module repo, create a build and include it into your Companion installation (there's an option for this for module development). I can give you detailed info if you want to try this way.

**Comment by Suenee** [comment]
> Unfortunately not really an easy way. You will have to clone the module repo, create a build and include it into your Companion installation (there's an option for this for module development). I can give you detailed info if you want to try this way.

Now they released COMPANION 3.4.4 but I do not see that your module was included. Why?

**Comment by fmalcher** [comment]
It's a patch version (3.4.z), new features are only added in minor versions (3.y.0).

**Comment by Suenee** [comment]
> It's a patch version (3.4.z), new features are only added in minor versions (3.y.0).

So I need to wait for 3.5 or higger?

**Comment by fmalcher** [comment]
That's right. Unfortunately, I do not know when this will be. But I will mergr this PR before the 3.5 release.
Once merged, the feature will already be available in the experimental/nightly releases of Companion.

**Comment by fmalcher** [comment]
@Suenee 3.5.0 is now available 😊 please let me know if everything works for you

**Comment by Suenee** [comment]
> [@Suenee](https://github.com/Suenee) 3.5.0 is now available 😊 please let me know if everything works for you

Hi, my friend!

I made some simple tests and I see it works well. Thank you so much!
